keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md

```diff
@@ -0,0 +1,337 @@
+
+#### Motivation for “pick the largest constraint” concept
+
+Take a look at the example:
+
+- `curReplicas = 10`
+- `calculatedReplicas = 20`
+
+The user specifies only one HPA parameter `ScaleUpPods = 5` and expects that number of replicas to be set to 15 during the next HPA controller loop.
+But the algorithm picks the largest change instead:
+
+    Constraint1 = 10 * 2 = 20 (as ScaleUpPercent = 100 by default)
+    Constraint2 = 10 + 5 = 15 (as ScaleUpPods = 5, set by the user)
+    scaleUpLimit = max(20, 15) = 20
+    desiredReplicas = 20
+
+The user might expect that the autoscaler would use the smallest constraint (15), not the largest one (20). This is not intuitive, but it does make sense if considered thoroughly.
+
+The main idea of the HPA is to autoscale because of a load increase to avoid request failures. It should work on both small cluster and large ones. For small clusters, the absolute number constraint works better (ScaleUpPods), for large clusters the percentage works better (ScaleUpPercentage).
+
+Example: If the current cluster size is `1` and calculated cluster size for this particular load is `20`, than we should reach it ASAP.
+
+For default values (ScaleUpPercent = 100, ScaleUpPods = 4) and “pick the largest constraint” concept, we’ll increase 1 -> 20 in 3 steps
+
+    1 -> 5 -> 10 -> 20
+
+The first step will use the “ScaleUpPods” limitation, next steps will use “ScaleUpPercent” limitation.
+
+In case of more intuitive “pick the smallest limit” concept, we’ll increase the cluster in 6 steps:
+
+    1 -> 2 -> 4 -> 8 -> 12 -> 16 -> 20
+
+Given that each steps takes [90 sec in worst case], we’ll respond to the load increase in `(6-3)*90 sec = ~ 5 min`.
+
+That’s too much, we should respond faster than that.
+
+[90 sec in worst case]: https://dzone.com/articles/kubernetes-autoscaling-101-cluster-autoscaler-hori-1
+
+#### Stabilization Window
+
+Stabilization window ([PR][], [RFC][]) is used to gather “scale-down-recommendations” during some time (default is 5min),
+and new number of replicas is set to the maximum of all recommendations.
+It will be applied after all the “scaleDown parameters” described above.
+We may want to control stabilization window size in addition to configuring the scale velocity,
+but at the moment, the “stabilization window feature” is considered to be an internal feature and is not exposed via any configuration.
+
+I’d suggest allowing to change the window size or turn it off if needed, but it should be discussed in a separate RFC.
+
+[PR]: https://github.com/kubernetes/kubernetes/pull/68122
+[RFC]: https://docs.google.com/document/d/1IdG3sqgCEaRV3urPLA29IDudCufD89RYCohfBPNeWIM/edit#heading=h.3tdw2jxiu42f
+
+
+#### API Changes
+
+The following API changes are needed:
+
+We should add `Scale{Up,Down}{Percent,Pods}` fields into the HPA spec
+
+The resulting solution will look like this:
+
+```golang
+type HorizontalPodAutoscalerScaleConstraints struct {
+    ScaleUpPercent Quantity
```

Comment:
We probably should replace Quantity with int32, as requested in https://github.com/kubernetes/kubernetes/pull/74525.